texastribune/salesforce-stripe
# Better customer ID handling

@x110dc Sorry if my Python sucks.

#### What's this PR do?
+ De-ajaxes Stripe card validation and instead reloads a pre-filled form with a prominent error message if a `CardError` is raised in Python.
+ Improves server-side form validation with more granular expectations for a field's value.

#### Why are we doing this? How does it help us?
So Stripe customer IDs are not visible on the client. Specifically:
+ No duplicate transactions from mistakenly resubmitting forms
+ No risk of someone gaining access to a customer ID and making a transaction on someone else's behalf

#### How should this be manually tested?
+ Build a fresh container with `make`
+ `make restart`
+ Fill out the form on `/donate`. Submit using either an error-raising test card number on [this list](https://stripe.com/docs/testing) or with a real card number. The page should reload with a red-boxed error atop the form. All fields besides card should be pre-filled with your previous input. The error should not disappear until you try to pay again.
+ Submit again, this time with a non-error-raising Stripe test card. The form should submit successfully, and you should see the thank-you page. Any referral or campaign IDs set as query params initially should have persisted through to this successful submit.
+ Repeat the last two steps on `/circleform`.
+ @x110dc Mind taking a glance at `forms.py` to ensure I didn't screw up anything glaringly obvious with the validation updates?

#### Have automated tests been added?
No. This app could use a front-end suite, but today is not that day.

#### Has this been tested on mobile?
Yes.

#### Are there performance implications?
No.

#### What are the relevant tickets?
Off-sprint.

#### How should this change be communicated to end users?
NA.

#### Next steps?
NA.

#### Smells?
NA.

#### TODOs:
* [x] Verify this JSON is sanitized properly to prevent XSS. @x110dc I might request your help here since you're tackling this anyway.
* [ ] After deployment, I'll do my usual slew of $1 test donations, including with Apple and Google Pay.

#### Has the relevant documentation/wiki been updated?
It will be before I merge.

#### Technical debt note
Even. Adds complexity because of pre-filling form state. Removes some because we're ditching the `create-customer` endpoint.
